ACAS-235: Create acasclient utility for validating CodeValues

bffrost opened the pull request:
## Description
I set out to make a utility that would validate that the codes in CodeValues are actually valid, i.e. they show up in the appropriate ACAS DDict. I quickly found that I'd implemented an early version of this, but it relied on an acasclient session being passed in at the time a CodeValue was instantiated, which basically never happens.

I added some unit tests, then started refactoring:
- *Big change*: Added CodeValue validation into a new `SimpleLsThing.validate` function that now happens as part of `save` unless explicitly opted-out.
- Created a new "DDict" generic interface, with implementations for ACAS DDicts and ACAS LsThing references. These are meant as a place to store fetching and checking logic for each `codeOrigin`.
- I then refactored the existing CodeValue validation logic to instead always go through DDict

## Related Issue
#42 

## How Has This Been Tested?
I've run the LsThing tests locally and they pass. I'm seeing some failures with the ACASClient unit tests that I'm not sure are related to my changes, and might be peculiar to my dev environment.

brianbolt reviewed (changes requested):
Overall I think it's great to add this type of business logic validation to ACAS. Something we are definitely lacking.  I have some nits and some worries about performance.  I also think test_009 could use some additionally comments on what exactly is being tested.

bffrost commented on requirements_dev.txt:
```diff
@@ -7,4 +7,5 @@ tox==3.14.0
 coverage==4.5.4
 Sphinx==1.8.5
 twine==1.14.0
+pandas==1.4.1
```
Thanks for the info. I was just trying to get my virtualenv for running unit tests working. I should've used setup.py. I'm going to revert these changes.

bffrost commented on tests/test_lsthing.py:
```diff
@@ -702,6 +717,80 @@ def test_007_advanced_search_interactions(self):
+        STATUS_DDICT = ACASDDict(PROJECT, STATUS)
+        proj_1.metadata[PROJECT_METADATA][STATUS_KEY] = CodeValue(ACTIVE, ddict=STATUS_DDICT)
+        proj_1.validate(self.client)
+        assert True
```
My idea here was just to confirm that we'd passed the "validate" function without throwing an error, but I think this is a silly way of doing it, so I'm going to remove these.

bffrost commented on tests/test_lsthing.py:
```diff
@@ -702,6 +717,80 @@ def test_007_advanced_search_interactions(self):
+        self.assertEqual(str(error), f"Invalid 'code':'{status_1}' provided for the given 'code_type':'{PROJECT}' and 'code_kind':'{STATUS}'")
+        # Now try adding a CodeValue that references an LsThing
+        proj_1.metadata[PROJECT_METADATA][STATUS_KEY] = CodeValue(ACTIVE, ddict=STATUS_DDICT)
+        proj_1.save(self.client)
```
Thanks. This line is actually just setup (saving a new LsThing) so that we can reference it below. I've adopted your suggestion above when I first validate `proj_1` then again below when I validate `proj_2`

bffrost commented on acasclient/lsthing.py:
```diff
@@ -2076,6 +2085,26 @@ def _upload_file_values_from_state_dict(state_dict):
+                if isinstance(value, CodeValue):
+                    if value.code:
+                        value.validate(client)
+        for _, values_dict in self.results.items():
```
Thanks, that's cleaner. Incorporated this idea.

bffrost commented on acasclient/lsthing.py:
```diff
@@ -2076,6 +2085,26 @@ def _upload_file_values_from_state_dict(state_dict):
+        :param client: Authenticated instance of acasclient.client
+        :type client: acasclient.client
+        """
+        for _, values_dict in self.metadata.items():
```
Thanks for the comment, I agree. I've shifted responsibility for fetching the contents of the DDicts up to some SimpleLsThing methods so that this can be done in bulk.

Also, `self.metadata` is the SimpleLsThing dictionary, so it won't have ignored or deleted values. The underlying `_ls_thing` object will, but we're not validating that.

bffrost commented on acasclient/lsthing.py:
```diff
@@ -2076,6 +2085,26 @@ def _upload_file_values_from_state_dict(state_dict):
         self.metadata = _upload_file_values_from_state_dict(self.metadata)
         self.results = _upload_file_values_from_state_dict(self.results)
+    
+    def validate(self, client):
```
I ended up removing client and pushing that responsibility onto the DDict.

bffrost commented on tests/test_lsthing.py:
```diff
@@ -702,6 +717,80 @@ def test_007_advanced_search_interactions(self):
                                        combine_terms_with_and=True)
         assert len(results) == 0
+    
+    def test_008_validate_codevalue(self):
```
Yep, added better comments here and below.

bffrost reviewed (commented):
@brianbolt Thank you for the feedback. I've made a number of changes to incorporate your comments. I'm still not feeling great about the approach of raising Exceptions on the first invalid codevalue - I think the more ACAS-canonical way would be to collect a series of error messages and attributes, then raise an error at the SimpleLsThing level in the final validation.

I'm going to work on that change, but pushing up my interim work so you can review.

iamanandkr commented on acasclient/lsthing.py:
```diff
@@ -971,24 +960,40 @@ def _validate_params(self, code, code_type, code_kind, code_origin, client):
         :type code_kind: str
         :param code_origin: Origin of DDictValue referenced, typically 'ACAS DDict'
         :type code_origin: str
```
Add documentation for  the new`ddict` argument?

iamanandkr commented on acasclient/lsthing.py:
```diff
@@ -971,24 +960,40 @@ def _validate_params(self, code, code_type, code_kind, code_origin, client):
+    def validate(self):
+        """Validate that this CodeValue conforms to the stored list of possible DDictValues in self.ddict
+
         :return: Error message, or None if valid
         :rtype: str | None
```
`ValueError` will be raised here in case of invalid ddict.

I believe we are going to remove the validation logic raising a `ValueError` exception, so this is optional based on what we end up doing.

iamanandkr commented on acasclient/lsthing.py:
```diff
@@ -971,24 +960,40 @@ def _validate_params(self, code, code_type, code_kind, code_origin, client):
-        if any([code == val_dict['code'] for val_dict in valid_val_dicts]):
+        if self.ddict:
+            self.ddict.check_value(self.code)
             return
```
Not required.

iamanandkr commented on acasclient/lsthing.py:
```diff
@@ -2076,6 +2109,62 @@ def _upload_file_values_from_state_dict(state_dict):
+                            ddict = ACASDDict(value.code_type, value.code_kind)
+                        elif value.code_origin.upper() == ACAS_LSTHING:
+                            ddict = ACASLsThingDDict(value.code_type, value.code_kind)
+                        ddicts[(ddict.code_type, ddict.code_kind, ddict.code_origin.upper())] = ddict
```
If the code_origin doesn't match ACAS_DDICT or ACAS_LSTHING, we will here raise an `AttributeError: NoneType object has no attribute ...`. Do you think we should raise a `ValueError` with custom message?

bffrost replied:
Yep, thanks.

iamanandkr commented on acasclient/lsthing.py:
```diff
@@ -2076,6 +2109,62 @@ def _upload_file_values_from_state_dict(state_dict):
+        ddicts = {}
+        state_dicts = [self.metadata, self.results]
+        for state_dict in state_dicts:
+            for _, values_dict in state_dict.items():
```
You could alternatively do:
```python
for values_dict in state_dict.values():
    ...
```

It could be the same with for-loops where we are ignoring the key.

iamanandkr commented on acasclient/ddict.py:
```diff
@@ -0,0 +1,56 @@
+class DDict(object):
+    """The DDict class is meant as a generic interface for any implementation of a Data Dictionary that
+    can be referenced by an ACAS CodeValue.
+    Any classes implementing this interface must implement the get_values() method.
```
What do you think about documenting that the caller should ensure `get_values` is called before calling `check_value`?

If the caller forgets to call `get_values` and directly calls `check_value`, they might get a `TypeError` because of `valid_values` being `None`.

bffrost replied:
Great suggestion, adding this both to the docs and adding an explicit exception if someone tries to call `check_value` without `valid_values` populated.

iamanandkr commented on acasclient/lsthing.py:
```diff
@@ -19,6 +21,8 @@
 
 
 ROW_NUM_KEY = 'row number'
+ACAS_DDICT = 'ACAS DDICT'
+ACAS_LSTHING = 'ACAS LSTHING'
```
I wonder if we should write this as: 🤔 
```python
ACAS_DDICT = ACASDDict.CODE_ORIGIN.upper()
ACAS_LSTHING = ACASLsThingDDict.CODE_ORIGIN.upper()
```

I was thinking on lines of removing duplication in knowledge.

brianbolt commented:
@bffrost going to let you work on the new changes and incorporating @iamanandkr feedback before the next review I do.

bffrost commented on acasclient/ddict.py:
```diff
@@ -0,0 +1,67 @@
+    CODE_ORIGIN = 'ACAS DDict'
+
+    def __init__(self, code_type, code_kind):
+        super(ACASDDict, self).__init__(code_type, code_kind, self.CODE_ORIGIN)
```
Yeah - I recall hitting this and realizing it is a python2 vs python3 issue. I think this way I have it is backwards-compatible with python2, so I'm not planning on changing it right now.

bffrost commented on acasclient/ddict.py:
```diff
@@ -0,0 +1,67 @@
+        self.code_origin = code_origin
+        self.valid_values = None
+
+    def get_values(self):
```
Agree on having `client` for consistency. I reworked a little how this gets populated so it is getting called from outside this class. In that case I'll rename it to `update_valid_values` since it is a public method.

bffrost commented on acasclient/ddict.py:
```diff
@@ -0,0 +1,67 @@
+        valid_codetables = client.get_ddict_values_by_type_and_kind(
+            self.code_type, self.code_kind)
+        self.valid_values = [val_dict['code'] for val_dict in valid_codetables]
+        if self.valid_values == []:
```
Thanks, I'll use the pythonic way.

bffrost commented:
Thank you @brianbolt  and @iamanandkr for the reviews. I've addressed your comments and also switched from an Exception-based approach to a `ValidationResult` based approach for handling errors/warnings in validation. If you could both review the new changes I'd really appreciate it!

iamanandkr commented on acasclient/ddict.py:
```diff
@@ -0,0 +1,63 @@
+    @validation_result
+    def check_value(self, value):
+        """Check if the value is within `self.valid_values` for the DDict."""
+        if not self.valid_values:
```
Curious: Do we want to show the custom message if `valid_values` is an empty list or only when `valid_values` is `None`? Checking if we should have this as,
```python
if self.valid_values is None:
    ...
```

bffrost replied:
Yeah, this maybe could be better but in the ACASDDict implementation of `update_valid_values` we throw a specific error if no valid values come back (empty list), so I think this will only be hit when someone calls without first calling `update_valid_values`. Going to merge this as-is.

iamanandkr commented on acasclient/lsthing.py:
```diff
@@ -971,23 +961,43 @@ def _validate_params(self, code, code_type, code_kind, code_origin, client):
+        """
+        if self.ddict:
+            return self.ddict.check_value(self.code)
+        else:
+            return True
```
Optional: We don't necessarily need it. If the function return value is `None` it will automatically be type cast to `ValidationResult(True, [])`.

iamanandkr reviewed (approved):
LGTM! I have added minor optional suggestions/comments, please feel free to ignore them.

brianbolt commented on acasclient/ddict.py:
```diff
@@ -36,18 +40,14 @@ class ACASDDict(DDict):
             self.code_type, self.code_kind)
         self.valid_values = [val_dict['code'] for val_dict in valid_codetables]
-        if self.valid_values == []:
+        if not self.valid_values:
```
Just noting that @iamanandkr comment about using `self.valid_values is None` applies here as well

bffrost replied:
Right, but here we are actually setting `self.valid_values` on the line above, so I think it will always be either empty list or populated list.

brianbolt reviewed (approved):
LGTM
